File: test/interval.c
typedef unsigned char  digital_t;
typedef unsigned short analog_t;
typedef unsigned long  timer_t;
typedef unsigned char  state_t;

enum {
    IN_interval_button,
    interval_NUM_INPUT
};

enum {
    OUT_interval_value,
    OUT_springback_value,
    OUT_alternate_value,
    interval_NUM_OUTPUT
};


enum {
    springback_off,
    springback_on
};

enum {
    alternate_low,
    alternate_high
};

typedef struct {
    digital_t input[interval_NUM_INPUT];
    digital_t output[interval_NUM_OUTPUT];
    timer_t   clk_alternate_Th;
    timer_t   clk_alternate_Tl;
    state_t  st_springback;
    state_t  st_alternate;
} interval_ctx_t;

extern int timer_init(timer_t* tp);
extern int timer_start(timer_t* tp);
extern int timer_stop(timer_t* tp);
extern int timer_timeout(timer_t* tp);

void init(interval_ctx_t* ctx)
{
    ctx->input[IN_interval_button] = 0;
    ctx->output[OUT_interval_value] = 0;
    timer_init(&ctx->clk_alternate_Th);
    timer_init(&ctx->clk_alternate_Tl);
    ctx->st_springback = springback_off;
    ctx->st_alternate = alternate_low;
}

void final(interval_ctx_t* ctx)
{
    timer_stop(&ctx->clk_alternate_Th);
    timer_stop(&ctx->clk_alternate_Tl);
}

void machine(interval_ctx_t* ctx)
{
    digital_t interval_button;
    digital_t springback_button;
    digital_t alternate_enable;

    interval_button = ctx->input[IN_interval_button];
    springback_button = ctx->input[IN_interval_button];
    switch(ctx->st_springback) {
    case springback_off:
        if (springback_button) {
            ctx->st_springback = springback_on;
            break;
        }
        break;
    case springback_on:
        if (springback_button) {
            ctx->st_springback = springback_off;
            break;
        }
        break;
    default: break;
    }
    ctx->output[OUT_springback_value] = (ctx->st_springback == springback_on);
    alternate_enable = ctx->output[OUT_springback_value];
    switch(ctx->st_alternate) {
    case alternate_high:
        if ((timer_timeout(&ctx->clk_alternate_Tl)) && (alternate_enable)) {
            ctx->st_alternate = alternate_low;
            timer_start(&ctx->clk_alternate_Th);
            break;
        }
        break;
    case alternate_low:
        if ((timer_timeout(&ctx->clk_alternate_Th)) && (alternate_enable)) {
            ctx->st_alternate = alternate_high;
            timer_start(&ctx->clk_alternate_Tl);
            break;
        }
        break;
    default: break;
    }
    ctx->output[OUT_alternate_value] = (ctx->st_alternate == alternate_high);
    ctx->output[OUT_interval_value] = (ctx->output[OUT_alternate_value]) && (ctx->output[OUT_springback_value]);
}

File: test/interval.erl
-module('interval').
-export([main/0,wait/3,loop/3,final/0]).
-define(BOOL(X), if (X) -> 1; true -> 0 end).


main() -> 
    wait(
    #{
        'springback_state' => 'off',
        'alternate_state' => 'low'    },
    #{
        'interval_button' => 0
    },
    #{
        'interval_value' => 0,
        'springback_value' => 0,
        'alternate_value' => 0
    }).

final() ->
    erlister_rt:stop_timer(clk_alternate_Th),
    erlister_rt:stop_timer(clk_alternate_Tl),
    ok.


wait(STATE,IN,OUT) ->
    receive
        {timeout,_Ref,_} -> loop(STATE,IN,OUT);
        INPUT -> loop(STATE,maps:merge(INPUT,IN),OUT)
    end.


loop(
    STATE = #{
                'springback_state' := ST_springback,
                'alternate_state' := ST_alternate    },
    IN = #{
        'interval_button' := I_interval_button
    },
    OUT = #{
        'interval_value' := OUT_interval_value,
        'springback_value' := OUT_springback_value,
        'alternate_value' := OUT_alternate_value
    }) ->
    IN_interval_button = I_interval_button,
    IN_springback_button = I_interval_button,
    ST1_springback = 
        case (ST_springback) of
        'off' ->
            if
                IN_springback_button ->
                    'on';
                true -> ST_springback
            end;
        'on' ->
            if
                IN_springback_button ->
                    'off';
                true -> ST_springback
            end;
        _ -> ST_springback
        end,
    OUT1_springback_value = ?BOOL((ST1_springback =:= 'on')),
    IN_alternate_enable = OUT_springback_value,
    CLK_alternate_Th = erlister_rt:read_timer(clk_interval_Th),
    CLK_alternate_Tl = erlister_rt:read_timer(clk_interval_Tl),
    ST1_alternate = 
        case (ST_alternate) of
        'high' ->
            if
                ((CLK_alternate_Tl =:= timeout)) andalso (IN_alternate_enable) ->
                    erlister_rt:timer_start(clk_alternate_Th,2000),
                    'low';
                true -> ST_alternate
            end;
        'low' ->
            if
                ((CLK_alternate_Th =:= timeout)) andalso (IN_alternate_enable) ->
                    erlister_rt:timer_start(clk_alternate_Tl,2000),
                    'high';
                true -> ST_alternate
            end;
        _ -> ST_alternate
        end,
    OUT1_alternate_value = ?BOOL((ST1_alternate =:= 'high')),
    OUT1_interval_value = ?BOOL((OUT_alternate_value) andalso (OUT_springback_value)),
    wait(STATE#{
                'springback_state' => ST1_springback,
                'alternate_state' => ST1_alternate    },
    IN,
    OUT#{
        'interval_value' => OUT1_interval_value,
        'springback_value' => OUT1_springback_value,
        'alternate_value' => OUT1_alternate_value
    }).
